src/auth/auth.middleware.ts

```diff
@@ -0,0 +1,145 @@
+CAR_FACTORY.codecs.add(DAG_JOSE)
+
+const VERIFIER = new DID({ resolver: KeyDIDResolver.getResolver() })
+
+export function parseAllowedDIDs(dids: string | undefined): Set<string> {
+  if (dids) {
+    const parts = dids.split(',')
+    return new Set(parts)
+  } else {
+    return new Set()
+  }
+}
+
+export function auth(opts: AuthOpts): Handler {
+  const hasAllowedDIDsList = opts.allowedDIDs.size > 0
+
+  /**
+   * @dev If the request has a did header, it means we have already confirmed the did
+   * is registered. If the request has no did, it means we have already
+   * confirmed the IP address making the request is on our allowlist. If the
+   * request contains a body, it means we have already verified the digest
+   * header can be trusted here.
+   * All of this logic mentioned above lives outside of this app.
+   * Notice that the absense of a did header or body bypasses any checks below
+   * this app will still work if the logice above is not in place.
+   */
+  return async function (req: Request, res: Response, next: NextFunction) {
+    const logger = opts.logger
+
+    // Use auth lambda
+    const didFromHeader = req.header('did')
+    if (didFromHeader && req.body) {
+      const digest = buildBodyDigest(req.header('Content-Type'), req.body)
+      if (req.header('digest') === digest) {
+        ServiceMetrics.count(METRIC_NAMES.AUTH_ALLOWED, 1, { did: didFromHeader })
+        return next()
+      } else {
+        logger?.verbose(`Disallowed: Auth lambda: Invalid digest`)
+        return disallow(res)
+      }
+    }
+
+    // Authorization Header
+    if (hasAllowedDIDsList) {
+      const authorizationHeader = req.header('Authorization') || ''
+      const bearerTokenMatch = AUTH_BEARER_REGEXP.exec(authorizationHeader)
+      const jws = bearerTokenMatch?.[1]
+      if (!jws) {
+        logger?.verbose(`Disallowed: No authorization header`)
+        return disallow(res)
+      }
+      const verifyJWSResult = await VERIFIER.verifyJWS(jws)
+      const did = verifyJWSResult.didResolutionResult.didDocument?.id
+      if (!did) {
+        logger?.verbose(`Disallowed: No DID`)
+        return disallow(res)
+      }
+      const nonce = verifyJWSResult.payload?.['nonce']
+      const digest = verifyJWSResult.payload?.['digest']
+      if (!nonce || !digest) {
+        logger?.verbose(`Disallowed: No nonce or No digest`)
+        return disallow(res)
+      }
+      if (!isAllowedDID(did, opts)) {
+        logger?.verbose(`Disallowed: ${did}`)
+        return disallow(res)
+      }
+
+      const body = req.body
+      const contentType = req.header('Content-Type')
+      const digestCalculated = buildBodyDigest(contentType, body)
+      const isCorrectDigest = digestCalculated == digest
+      if (!isCorrectDigest) {
+        logger?.verbose(`Disallowed: Incorrect digest for DID ${did}`)
+        return disallow(res)
+      }
+      ServiceMetrics.count(METRIC_NAMES.AUTH_ALLOWED, 1, { did: did })
```

Comment:
Won't that double count successes?

Reply:
I mean, since the same request could be counted at L54

Reply:
That would not double the counter. L54 is immediately followed by `return next()` terminal, so L96 is not going to be reached.

https://github.com/ceramicnetwork/ceramic-anchor-service/pull/1234/files/9a0ea8636ca8b61d473ccd1126602f1ec72a2bfa#diff-56f006a2ae9c4ee0e62a81ef1d1a0fd7244216124906e7b330a1d3d33231402dR55